### PR TITLE
[Feature] Dynamic source sink

### PR DESCRIPTION
**Pull request by mproch**
This PR allows for defining dynamic parameters with GenericNodeTransformer for sources and sinks.
Some restrictions:
- it makes no sense to change ValidationContext in sinks
- for sources there are some restrictions for ValidationContext (e.g. we allow only 'input' variable in Flink sources...)

Note: this PR introduces tests and samples, but the only substantial changes to production code are in ProcessCompiler

### CODE REVIEW

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala [hunk 1]
```diff
@@ -24,7 +24,6 @@ case class FullNameNotCompatible(var first: CharSequence, var middle: CharSequen
       case 0 => first = value.asInstanceOf[CharSequence]
       case 1 => middle = value.asInstanceOf[CharSequence]
       case 2 => last = value.asInstanceOf[CharSequence]
-      case 2 => sex = value.asInstanceOf[CharSequence]
```
Comment (lciolecki) [inline engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala:27]: Why did you remove it?

Comment (mproch) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala:27]: I try to remove compiler warnings when it's possible :)

File: engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/helpers/SampleNodes.scala [hunk 2]
```diff
@@ -480,6 +484,88 @@ object SampleNodes {
...
+    override type State = Nothing
+
+    override def contextTransformation(context: ValidationContext, dependencies: List[NodeDependencyValue])(implicit nodeId: NodeId)
+      : this.NodeTransformationDefinition = {
+      case TransformationStep(Nil, _) => NextParameters(initialParameters)
+      case TransformationStep(("type", DefinedEagerParameter(value: String, _))::Nil, None) =>
+        val versions = value match {
+          case "type1" => List(1, 2)
+          case "type2" => List(3, 4)
+          case _ => ???
```
Comment (lciolecki) [inline engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/helpers/SampleNodes.scala:498]: Please add comment that this is only for example.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala [hunk 3]
```diff
@@ -73,16 +77,53 @@ class CustomNodeValidator(modelData: ModelData) extends NodeDataValidator[Custom
...
+    }
+  }
+
+}
+
+class CustomNodeValidator(val modelData: ModelData) extends WithParametersNodeValidator[CustomNode] {
+
   override def validate(nodeData: CustomNode, validationContext: ValidationContext)(implicit metaData: MetaData): ValidationResponse = {
     val transformer = modelData.processWithObjectsDefinition.customStreamTransformers(nodeData.nodeType)._1.obj
     transformer match {
```
Comment (lciolecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala:97]: Thise fragment we can extract to some mixin and reuse it in SinkNodeValidator and SourceNodeValidator.

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala (hunk 1 above)
Comment (lciolecki) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala:27]: Okey, but field in class still exists? What kinf of warning was there?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala (hunk 3 above)
Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala]: I extracted a bit

File: engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/helpers/SampleNodes.scala (hunk 2 above)
Comment (mproch) [reply engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/helpers/SampleNodes.scala:499]: done

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala (hunk 1 above)
Comment (mproch) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala:27]: ok, I understand now :)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaDynamicAvroSinkFactory.scala [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package pl.touk.nussknacker.engine.avro.sink
+
+class KafkaDynamicAvroSinkFactory {
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaDynamicAvroSinkFactory.scala:3]: What it does here?

Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaDynamicAvroSinkFactory.scala]: good catch ;)
